### PR TITLE
Block Storage Checkpointing

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
Title says it all.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-652

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: block-storage/src/test/scala/coop/rchain/blockstorage/BlockGen.scala [hunk 1]
```diff
@@ -18,26 +18,22 @@ object BlockGen {
...
+    for {
+      byteArray <- listOfN(size, arbitrary[Byte])
+    } yield ByteString.copyFrom(byteArray.toArray)
 
   val blockElementGen: Gen[BlockMessage] =
     for {
-      hash      <- arbitrary[BlockHash]
-      validator <- arbitrary[Validator]
+      hash      <- byteStringGen(32)
+      validator <- byteStringGen(32)
```
Comment (KentShikama) [inline block-storage/src/test/scala/coop/rchain/blockstorage/BlockGen.scala]: Might duplicate https://github.com/rchain/rchain/pull/2129 but I guess it is fine for now.

File: block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala [hunk 2]
```diff
@@ -242,70 +299,67 @@ object FileLMDBIndexBlockStore {
...
-    } yield directoriesList
+    } yield filesList
 
-  private def listDirectories[F[_]: Sync](dirPath: Path): StorageErrT[F, List[Path]] = {
+  private def listFiles[F[_]: Sync](dirPath: Path): StorageErrT[F, List[Path]] = {
     type TEMP[A] = StorageErrT[F, A]
     for {
       inDirectoryList <- listInDirectory(dirPath)
-      directoryList   <- inDirectoryList.filterA[TEMP](isDirectory)
+      directoryList   <- inDirectoryList.filterA[TEMP](isRegularFile)
```
Comment (KentShikama) [inline block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala]: What makes this TEMP?

Comment (itegulov) [reply block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala]: Accidently left this in in the last PR while playing with scalac

File: block-storage/src/test/scala/coop/rchain/blockstorage/BlockStoreTest.scala [hunk 3]
```diff
@@ -106,8 +99,8 @@ trait BlockStoreTest
     withStore { store =>
       val exception = new RuntimeException("msg")
 
-      def elem = {
-        blockHashElementsGen.sample.get
+      def elem: (BlockHash, BlockMessage) = {
+        blockElementGen.sample.get
```
Comment (lukasz-golebiewski) [inline block-storage/src/test/scala/coop/rchain/blockstorage/BlockStoreTest.scala]: Why is this line needed here?

File: block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala [hunk 4]
```diff
@@ -23,17 +23,28 @@ import org.lmdbjava.DbiFlags.MDB_CREATE
...
 
 class FileLMDBIndexBlockStore[F[_]: Monad: Sync: Log] private (
     lock: Semaphore[F],
     env: Env[ByteBuffer],
     index: Dbi[ByteBuffer],
+    storagePath: Path,
+    checkpointsDir: Path,
     blockMessageRandomAccessFileRef: Ref[F, RandomAccessFile],
-    checkpointsRef: Ref[F, List[Checkpoint]]
+    checkpointsRef: Ref[F, Map[Int, Checkpoint]],
```
Comment (lukasz-golebiewski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala]: It looks to me that all those `Ref`s are used within `lock.withPermit` blocks. Since this prevents races, it would make more sense to put all of them inside a single `MonadState` object

**Comment by itegulov** [comment]
bors try

**Comment by itegulov** [comment]
@lukasz-golebiewski PTAL
